#### Import relevant libraries

In [1]:
import numpy as np
import pandas as pd

#### Load the data

In [2]:
raw_data = pd.read_csv('C:/Users/johni/OneDrive/Documents/Studies/Data Analytics Cert/CIND 820/Code/raw_data.csv')

C:\Users\johni\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (22) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [3]:
raw_data.head()

url       region  \
0  https://abilene.craigslist.org/cto/d/aspermont...  abilene, TX   
1  https://abilene.craigslist.org/ctd/d/liberty-h...  abilene, TX   
2  https://abilene.craigslist.org/ctd/d/liberty-h...  abilene, TX   
3  https://abilene.craigslist.org/cto/d/merkel-20...  abilene, TX   
4  https://abilene.craigslist.org/cto/d/breckenri...  abilene, TX   

                       region_url  price  year manufacturer         model  \
0  https://abilene.craigslist.org   9000  2009    chevrolet  suburban lt2   
1  https://abilene.craigslist.org  31999  2012          ram          2500   
2  https://abilene.craigslist.org  16990  2003          ram          3500   
3  https://abilene.craigslist.org   6000  2002          gmc   sierra 1500   
4  https://abilene.craigslist.org  37000  2012    chevrolet          3500   

   condition    cylinders    fuel  ...                VIN drive       size  \
0       good  8 cylinders     gas  ...  1GFNC26099R211276   rwd  full-size   
1        NaN          NaN  diesel  ...                NaN   NaN        NaN   
2        NaN          NaN  diesel  ...                NaN   NaN        NaN   
3       good  8 cylinders     gas  ...                NaN   4wd        NaN   
4  excellent  8 cylinders  diesel  ...                NaN   4wd  full-size   

     type paint_color                                          image_url  \
0     SUV       white  https://images.craigslist.org/00909_5P6f1E6zHl...   
1     NaN         NaN  https://images.craigslist.org/00u0u_24XW8uhLYP...   
2     NaN         NaN  https://images.craigslist.org/00l0l_H1UKTJ4sTI...   
3  pickup       white  https://images.craigslist.org/00M0M_2wKzc7dKLU...   
4  pickup      silver  https://images.craigslist.org/00r0r_eW1YWlKfCG...   

                                         description        lat        long  \
0  2WD 1/2 ton\nLeather Captains Chairs\nIn good ...  33.130100 -100.234400   
1  www.GETADIESEL.com\n2012 RAM 2500 SLT 4X4 MEGA...  30.648368  -97.862949   
2  www.GETADIESEL.com\n2003 DODGE RAM 3500 LARAMI...  30.648516  -97.862434   
3  2002 GMC Sierra Extended Cab Truck For Sale! R...  32.444400  -99.992400   
4  2012 Chevrolet 3500 178k miles, runs and drive...  32.781689  -98.942248   

  posting_date  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  

[5 rows x 23 columns]

In [4]:
raw_data.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1961218 entries, 0 to 1961217
Data columns (total 23 columns):
 #   Column        Non-Null Count    Dtype  
---  ------        --------------    -----  
 0   url           1961218 non-null  object 
 1   region        1961218 non-null  object 
 2   region_url    1961218 non-null  object 
 3   price         1961218 non-null  int64  
 4   year          1961218 non-null  int64  
 5   manufacturer  1873885 non-null  object 
 6   model         1932807 non-null  object 
 7   condition     1098169 non-null  object 
 8   cylinders     1183030 non-null  object 
 9   fuel          1945959 non-null  object 
 10  odometer      1692151 non-null  float64
 11  title_status  1943577 non-null  object 
 12  transmission  1947925 non-null  object 
 13  VIN           1153230 non-null  object 
 14  drive         1379288 non-null  object 
 15  size          615960 non-null   object 
 16  type          1447537 non-null  object 
 17  paint_color   1338645 non-n

#### Check the descriptive statistics of the data

In [5]:
pd.set_option('display.float_format', lambda x: '%.0f' % x)
raw_data.describe().round(0)

price    year   odometer     lat    long
count    1961218 1961218    1692151 1923237 1923237
mean       68799    2005     101178      38     -94
std     10856876     103    1583663       6      18
min            0       0          0     -84    -178
25%         4399    2007      44372      35    -109
50%        10000    2012      91500      39     -88
75%        19995    2016     136770      42     -81
max   3787876818    2022 2043755555      84     174

#### Preprocessing 1 (P1) = Remove the following columns: 
'url' - does not affect the price

'region_url' - because web address of the region where it being sold does not affect the price

'VIN' - the vehicle identification number is a unique code that is used to identify where and when the vehicle was made. This does not affect price.

'lat' and 'long' - because location is already indicated by the region

'image_url' and 'description' - It was planned to transform this into a binary variable indicating whether this information was provided. The model would have determined whether having these information would affect the price or not. However, approximately all the rows have this information making these binary variables inconsequential to the model


In [6]:
raw_data_p1 = raw_data.drop(['url','region_url','VIN','lat','long','image_url', 'description'],axis=1)

In [7]:
raw_data_p1.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1961218 entries, 0 to 1961217
Data columns (total 16 columns):
 #   Column        Non-Null Count    Dtype  
---  ------        --------------    -----  
 0   region        1961218 non-null  object 
 1   price         1961218 non-null  int64  
 2   year          1961218 non-null  int64  
 3   manufacturer  1873885 non-null  object 
 4   model         1932807 non-null  object 
 5   condition     1098169 non-null  object 
 6   cylinders     1183030 non-null  object 
 7   fuel          1945959 non-null  object 
 8   odometer      1692151 non-null  float64
 9   title_status  1943577 non-null  object 
 10  transmission  1947925 non-null  object 
 11  drive         1379288 non-null  object 
 12  size          615960 non-null   object 
 13  type          1447537 non-null  object 
 14  paint_color   1338645 non-null  object 
 15  posting_date  884997 non-null   object 
dtypes: float64(1), int64(2), object(13)
memory usage: 239.4+ MB


#### P2 = Move price to the first column

In [8]:
raw_data_p2 = raw_data_p1.copy()
price_column = raw_data_p2.pop('price')
raw_data_p2.insert(0, 'price', price_column)
raw_data_p2.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1961218 entries, 0 to 1961217
Data columns (total 16 columns):
 #   Column        Non-Null Count    Dtype  
---  ------        --------------    -----  
 0   price         1961218 non-null  int64  
 1   region        1961218 non-null  object 
 2   year          1961218 non-null  int64  
 3   manufacturer  1873885 non-null  object 
 4   model         1932807 non-null  object 
 5   condition     1098169 non-null  object 
 6   cylinders     1183030 non-null  object 
 7   fuel          1945959 non-null  object 
 8   odometer      1692151 non-null  float64
 9   title_status  1943577 non-null  object 
 10  transmission  1947925 non-null  object 
 11  drive         1379288 non-null  object 
 12  size          615960 non-null   object 
 13  type          1447537 non-null  object 
 14  paint_color   1338645 non-null  object 
 15  posting_date  884997 non-null   object 
dtypes: float64(1), int64(2), object(13)
memory usage: 239.4+ MB


#### P3 = Remove row with missing data in any of the independent variable

In [9]:
#Get the list of independent variables
independent_variables = raw_data_p2.iloc[:,1:-1].columns.tolist()
independent_variables

['region',
 'year',
 'manufacturer',
 'model',
 'condition',
 'cylinders',
 'fuel',
 'odometer',
 'title_status',
 'transmission',
 'drive',
 'size',
 'type',
 'paint_color']

In [10]:
raw_data_p3 =raw_data_p2.dropna(how='any', subset=independent_variables)
raw_data_p3.shape

(389708, 16)

In [11]:
#get the number of missing values per column and the percentage to the total number of records
print("column      number of missing values      percentage")
column_names = list(raw_data_p3.columns)
for c in column_names:
    num_missing_values = sum(raw_data_p3[c].isnull())
    percentage = round(num_missing_values/raw_data_p3.shape[0]*100, 2)
    print ("%-15s   %10s           %10.2f" % (c, f"{num_missing_values:,}", percentage))

column      number of missing values      percentage
price                      0                 0.00
region                     0                 0.00
year                       0                 0.00
manufacturer               0                 0.00
model                      0                 0.00
condition                  0                 0.00
cylinders                  0                 0.00
fuel                       0                 0.00
odometer                   0                 0.00
title_status               0                 0.00
transmission               0                 0.00
drive                      0                 0.00
size                       0                 0.00
type                       0                 0.00
paint_color                0                 0.00
posting_date         222,919                57.20


#### P4 = Clean the 'model' column by changing all values to lower caps and removing white spaces

In [12]:
#Check the number of unique values for 'model'
raw_data_p4 = raw_data_p3.copy()
raw_data_p3['model'].nunique()

25434

In [13]:
#make everything in the model lowercase
raw_data_p4['model'] = raw_data_p4['model'].str.lower()

In [14]:
#remove white spaces from column 'model'
raw_data_p4['model'] = raw_data_p4['model'].str.strip()

#### P5 = Remove the outliers in price and odometer by using the IQR rule

In [15]:
#Determine the first quartile and third quartile and compute the inter-quartile range 
Q1 = raw_data_p4.quantile(0.25)
Q3 = raw_data_p4.quantile(0.75)
IQR = Q3 - Q1
print(IQR)

price      11800
year           9
odometer   83800
dtype: float64


In [16]:
print(Q1)
print(Q3)

price       4100
year        2005
odometer   70000
Name: 0.25, dtype: float64
price       15900
year         2014
odometer   153800
Name: 0.75, dtype: float64


In [17]:
#Remove the outliers from the dataset and check how may rows are left
raw_data_p5 = raw_data_p4[~((raw_data_p4 < (Q1 - 1.5 * IQR))|(raw_data_p4 > (Q3 + 10 * IQR))).any(axis=1)]
raw_data_p5.shape

(376458, 16)

In [18]:
#Number of rows removed in P5
print("Number of rows removed in P5: ", raw_data_p4.shape[0]-raw_data_p5.shape[0])
#percentage of rows of original dataset removed
print("Percentage: ", round(((raw_data_p4.shape[0]-raw_data_p5.shape[0])/raw_data.shape[0]*100),2))
raw_data_p5.info()

Number of rows removed in P5:  13250
Percentage:  0.68
<class 'pandas.core.frame.DataFrame'>
Int64Index: 376458 entries, 0 to 1961174
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   price         376458 non-null  int64  
 1   region        376458 non-null  object 
 2   year          376458 non-null  int64  
 3   manufacturer  376458 non-null  object 
 4   model         376458 non-null  object 
 5   condition     376458 non-null  object 
 6   cylinders     376458 non-null  object 
 7   fuel          376458 non-null  object 
 8   odometer      376458 non-null  float64
 9   title_status  376458 non-null  object 
 10  transmission  376458 non-null  object 
 11  drive         376458 non-null  object 
 12  size          376458 non-null  object 
 13  type          376458 non-null  object 
 14  paint_color   376458 non-null  object 
 15  posting_date  160379 non-null  object 
dtypes: float64(1), int64(2), object(13)


In [19]:
raw_data_p5.describe()

price   year  odometer
count 376458 376458    376458
mean   11589   2010    115285
std    10777      6     63939
min        0   1992         0
25%     4200   2006     71200
50%     8399   2010    112500
75%    15900   2014    154000
max   131500   2022    991152

#### P6 = Drop rows that have a odometer readings of more than 372,000

Professional mechanics say that 12,000 miles per year is an accurate estimate of a car that has not been overdriven. The cars in the dataset have years from 1992 to 2022 which is 31 years. Therefore, the maximum odometer reading expected for the dataset should be 372,000 miles. 

In [20]:
raw_data_p6 = raw_data_p5[raw_data_p5['odometer']<372000]
raw_data_p6.shape

(375801, 16)

In [21]:
#number of rows removed in P6
print("Number of rows removed in P6: ", raw_data_p5.shape[0]-raw_data_p6.shape[0])
#percentage of rows of original dataset removed
print("Percentage: ", round(((raw_data_p5.shape[0]-raw_data_p6.shape[0])/raw_data.shape[0]*100),2))
raw_data_p6.info()

Number of rows removed in P6:  657
Percentage:  0.03
<class 'pandas.core.frame.DataFrame'>
Int64Index: 375801 entries, 0 to 1961174
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   price         375801 non-null  int64  
 1   region        375801 non-null  object 
 2   year          375801 non-null  int64  
 3   manufacturer  375801 non-null  object 
 4   model         375801 non-null  object 
 5   condition     375801 non-null  object 
 6   cylinders     375801 non-null  object 
 7   fuel          375801 non-null  object 
 8   odometer      375801 non-null  float64
 9   title_status  375801 non-null  object 
 10  transmission  375801 non-null  object 
 11  drive         375801 non-null  object 
 12  size          375801 non-null  object 
 13  type          375801 non-null  object 
 14  paint_color   375801 non-null  object 
 15  posting_date  160120 non-null  object 
dtypes: float64(1), int64(2), object(13)
me

#### P7 = Remove cars with price less than 500

In [22]:
raw_data_p7 = raw_data_p6[raw_data_p6['price']>500]
raw_data_p7.shape

(350884, 16)

In [23]:
#number of rows removed in P7
print("Number of rows removed in P7: ", raw_data_p6.shape[0]-raw_data_p7.shape[0])
#percentage of rows of original dataset removed
print("Percentage: ", round(((raw_data_p6.shape[0]-raw_data_p7.shape[0])/raw_data.shape[0]*100),2))
raw_data_p7.info()

Number of rows removed in P7:  24917
Percentage:  1.27
<class 'pandas.core.frame.DataFrame'>
Int64Index: 350884 entries, 0 to 1961171
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   price         350884 non-null  int64  
 1   region        350884 non-null  object 
 2   year          350884 non-null  int64  
 3   manufacturer  350884 non-null  object 
 4   model         350884 non-null  object 
 5   condition     350884 non-null  object 
 6   cylinders     350884 non-null  object 
 7   fuel          350884 non-null  object 
 8   odometer      350884 non-null  float64
 9   title_status  350884 non-null  object 
 10  transmission  350884 non-null  object 
 11  drive         350884 non-null  object 
 12  size          350884 non-null  object 
 13  type          350884 non-null  object 
 14  paint_color   350884 non-null  object 
 15  posting_date  150986 non-null  object 
dtypes: float64(1), int64(2), object(13)


#### P8 = Remove the records with a car model that has a frequency of less than 150

In [24]:
#Count the number of unique models in the dataset
raw_data_p7['model'].nunique()

22577

In [25]:
#Count the number of unique models with a frequency greater than 150
model_counts = pd.DataFrame(raw_data_p7['model'].value_counts())
model_counts[model_counts['model']>150]

model
f-150                  7826
silverado 1500         6323
1500                   4740
silverado              2869
accord                 2775
...                     ...
wrangler sahara         153
cc4500                  152
tribute                 152
outback 2.5i limited    151
xc60                    151

[377 rows x 1 columns]

In [26]:
#Percentage of models with a frequency greater than 150
model_counts[model_counts['model']>150].count()[0]/raw_data_p7['model'].nunique()*100

1.6698409886167338

In [27]:
#Percentage of the records that have models that have a frequency greater than 150
model_counts[model_counts['model']>150].sum()[0]/raw_data_p7.shape[0]*100

60.771081041027806

This shows that 377 models represent 60.77% of the cars in the dataset. The number of other models with a frequency of less than 150 is 22,200. They represent the remaining 39.23% of the cars in the dataset.  

In [28]:
main_models = model_counts[model_counts['model']>150].index.tolist()
main_models

['f-150',
 'silverado 1500',
 '1500',
 'silverado',
 'accord',
 '2500',
 'camry',
 'silverado 2500hd',
 'sierra 1500',
 'civic',
 'tacoma',
 'escape',
 'altima',
 'wrangler',
 'explorer',
 'tahoe',
 'grand cherokee',
 'corolla',
 'mustang',
 'focus',
 'impala',
 'f150',
 'grand caravan',
 'equinox',
 '3500',
 'tundra',
 'prius',
 'cr-v',
 'malibu',
 'jetta',
 'suburban',
 'rav4',
 'fusion',
 'forester',
 'sentra',
 'sonata',
 'outback',
 'sierra 2500hd',
 'elantra',
 '4runner',
 'odyssey',
 'sierra',
 'rogue',
 'expedition',
 'cruze',
 'camry le',
 'fusion se',
 'pilot',
 'f-250 sd',
 'highlander',
 'ranger',
 'focus se',
 'f-250',
 'edge',
 'f250',
 'sienna',
 'camaro',
 'passat',
 'town & country',
 'corvette',
 'liberty',
 'f250 super duty',
 'taurus',
 'charger',
 'traverse',
 'colorado',
 'santa fe',
 'escalade',
 'civic lx',
 '3',
 'acadia',
 'yukon',
 'econoline',
 'wrangler unlimited',
 'express cargo van',
 'soul',
 'durango',
 'trailblazer',
 'cherokee',
 'dakota',
 'impreza'

In [29]:
raw_data_p8 = raw_data_p7.copy()
raw_data_p8 = raw_data_p8[raw_data_p8['model'].isin(main_models)]
raw_data_p8.shape

(213236, 16)

In [30]:
#number of rows removed in P8
print("Number of rows removed in P8: ", raw_data_p7.shape[0]-raw_data_p8.shape[0])
#percentage of rows of original dataset removed
print("Percentage: ", round(((raw_data_p7.shape[0]-raw_data_p8.shape[0])/raw_data.shape[0]*100),2))
raw_data_p8.info()

Number of rows removed in P8:  137648
Percentage:  7.02
<class 'pandas.core.frame.DataFrame'>
Int64Index: 213236 entries, 4 to 1961146
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   price         213236 non-null  int64  
 1   region        213236 non-null  object 
 2   year          213236 non-null  int64  
 3   manufacturer  213236 non-null  object 
 4   model         213236 non-null  object 
 5   condition     213236 non-null  object 
 6   cylinders     213236 non-null  object 
 7   fuel          213236 non-null  object 
 8   odometer      213236 non-null  float64
 9   title_status  213236 non-null  object 
 10  transmission  213236 non-null  object 
 11  drive         213236 non-null  object 
 12  size          213236 non-null  object 
 13  type          213236 non-null  object 
 14  paint_color   213236 non-null  object 
 15  posting_date  90637 non-null   object 
dtypes: float64(1), int64(2), object(13)

#### P9 = Remove the records with a region that has a frequency of less than 150

In [31]:
#Check the number of unique regions
raw_data_p8['region'].nunique()

628

In [32]:
#Check the number of regions with a frequency greater than 150
region_counts = pd.DataFrame(raw_data_p8['region'].value_counts())
region_counts[region_counts['region']>150]

region
vermont                 2837
indianapolis            1741
kansas city, MO         1711
detroit metro           1694
appleton-oshkosh-FDL    1684
...                      ...
springfield, IL          159
amarillo                 159
chico                    155
la crosse                154
huntington-ashland       153

[353 rows x 1 columns]

In [33]:
main_regions = region_counts[region_counts['region']>150].index.tolist()
main_regions

['vermont',
 'indianapolis',
 'kansas city, MO',
 'detroit metro',
 'appleton-oshkosh-FDL',
 'minneapolis / st paul',
 'boston',
 'worcester / central MA',
 'oklahoma city',
 'raleigh / durham / CH',
 'maine',
 'greenville / upstate',
 'new hampshire',
 'central NJ',
 'rhode island',
 'albuquerque',
 'western massachusetts',
 'louisville, KY',
 'omaha / council bluffs',
 'south jersey',
 'albany, NY',
 'philadelphia',
 'sarasota-bradenton',
 'anchorage / mat-su',
 'tampa bay area',
 'reno / tahoe',
 'new york city',
 'south bend / michiana',
 'north jersey',
 'dayton / springfield',
 'dallas / fort worth',
 'washington, DC',
 'san diego',
 'st louis, MO',
 'springfield, MO',
 'ventura county',
 'chicago',
 'rochester, NY',
 'norfolk / hampton roads',
 'fargo / moorhead',
 'colorado springs',
 'las vegas',
 'ft myers / SW florida',
 'san antonio',
 'green bay, WI',
 'tulsa, OK',
 'akron / canton',
 'syracuse, NY',
 'SF bay area',
 'hawaii',
 'madison, WI',
 'rochester',
 'louisville',
 

In [34]:
#Percentage of regions with a frequency greater than 150
region_counts[region_counts['region']>150].count()[0]/raw_data_p8['region'].nunique()*100

56.21019108280255

In [35]:
#Percentage of the records that have regions that have a frequency greater than 150
region_counts[region_counts['region']>150].sum()[0]/raw_data_p8.shape[0]*100

91.38278714663566

This means that 353 of the regions represent 91.38% of the data. The other 275 regions only represent 8.62% of the cars in the dataset. 

In [36]:
raw_data_p9 = raw_data_p8.copy()
raw_data_p9 = raw_data_p9[raw_data_p9['region'].isin(main_regions)]
raw_data_p9.shape

(194861, 16)

In [37]:
#number of rows removed in P9
print("Number of rows removed in P9: ", raw_data_p8.shape[0]-raw_data_p9.shape[0])
#percentage of rows of original dataset removed
print("Percentage: ", round(((raw_data_p8.shape[0]-raw_data_p9.shape[0])/raw_data.shape[0]*100),2))
raw_data_p9.info()

Number of rows removed in P9:  18375
Percentage:  0.94
<class 'pandas.core.frame.DataFrame'>
Int64Index: 194861 entries, 102 to 1961146
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   price         194861 non-null  int64  
 1   region        194861 non-null  object 
 2   year          194861 non-null  int64  
 3   manufacturer  194861 non-null  object 
 4   model         194861 non-null  object 
 5   condition     194861 non-null  object 
 6   cylinders     194861 non-null  object 
 7   fuel          194861 non-null  object 
 8   odometer      194861 non-null  float64
 9   title_status  194861 non-null  object 
 10  transmission  194861 non-null  object 
 11  drive         194861 non-null  object 
 12  size          194861 non-null  object 
 13  type          194861 non-null  object 
 14  paint_color   194861 non-null  object 
 15  posting_date  81409 non-null   object 
dtypes: float64(1), int64(2), object(13

#### P10 = Convert year to string

In [38]:
raw_data_p10 = raw_data_p9.copy()
raw_data_p10['year'] = raw_data_p10['year'].astype(str)
raw_data_p10.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 194861 entries, 102 to 1961146
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   price         194861 non-null  int64  
 1   region        194861 non-null  object 
 2   year          194861 non-null  object 
 3   manufacturer  194861 non-null  object 
 4   model         194861 non-null  object 
 5   condition     194861 non-null  object 
 6   cylinders     194861 non-null  object 
 7   fuel          194861 non-null  object 
 8   odometer      194861 non-null  float64
 9   title_status  194861 non-null  object 
 10  transmission  194861 non-null  object 
 11  drive         194861 non-null  object 
 12  size          194861 non-null  object 
 13  type          194861 non-null  object 
 14  paint_color   194861 non-null  object 
 15  posting_date  81409 non-null   object 
dtypes: float64(1), int64(1), object(14)
memory usage: 25.3+ MB


In [60]:
#Export dataset with numerical variables to a csv file for exploratory data analysis
raw_data_p10.to_csv('preprocessed_data_eda_num.csv', index=False)

#### P11 = Convert odometer into a categorical variable

In [39]:
#check the minimum and maximum values
raw_data_p10.describe()

price  odometer
count 194861    194861
mean   12277    118468
std    10331     59912
min      505         0
25%     4995     77082
50%     8995    116424
75%    16500    156793
max   131500    371000

In [40]:
#create a list of bins for the ranges
odo_bins = []
for i in range(32):
    odo_bins.append(12000*i)
odo_bins

[0,
 12000,
 24000,
 36000,
 48000,
 60000,
 72000,
 84000,
 96000,
 108000,
 120000,
 132000,
 144000,
 156000,
 168000,
 180000,
 192000,
 204000,
 216000,
 228000,
 240000,
 252000,
 264000,
 276000,
 288000,
 300000,
 312000,
 324000,
 336000,
 348000,
 360000,
 372000]

In [41]:
#create a list of labels for the ranges
odo_labels = ['0-12k', '12k-24k', '24k-36k', '36k-48k', '48k-60k', '60k-72k', '72k-84k', '84k-96k', '96k-108k', '108k-120k',
            '120k-132k', '132k-144k', '144k-156k', '156k-168k', '168k-180k', '180k-192k', '192k-204k', '204k-216k',
            '216k-228k', '228k-240k', '240k-252k', '252k-264k', '264k-276k', '276k-288k', '288k-300k', '300k-312k',
            '312k-324k', '324k-336k', '336k-348k', '348k-360k', '360k-372k']

In [42]:
#Create the new column 'odometer_range'
raw_data_p11 = raw_data_p10.copy()
odo_category = pd.cut(raw_data_p11.odometer,bins=odo_bins,labels=odo_labels)
raw_data_p11.insert(9,'odometer_range',odo_category)
raw_data_p11.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 194861 entries, 102 to 1961146
Data columns (total 17 columns):
 #   Column          Non-Null Count   Dtype   
---  ------          --------------   -----   
 0   price           194861 non-null  int64   
 1   region          194861 non-null  object  
 2   year            194861 non-null  object  
 3   manufacturer    194861 non-null  object  
 4   model           194861 non-null  object  
 5   condition       194861 non-null  object  
 6   cylinders       194861 non-null  object  
 7   fuel            194861 non-null  object  
 8   odometer        194861 non-null  float64 
 9   odometer_range  194343 non-null  category
 10  title_status    194861 non-null  object  
 11  transmission    194861 non-null  object  
 12  drive           194861 non-null  object  
 13  size            194861 non-null  object  
 14  type            194861 non-null  object  
 15  paint_color     194861 non-null  object  
 16  posting_date    81409 non-null   ob

In [43]:
"""The above table shows that 'odometer_range' has some null values. We check the rows which
have null values in this column."""
na_odo_range = raw_data_p11[raw_data_p11['odometer_range'].isna()]
na_odo_range

price               region  year manufacturer                  model  \
6485      7400         columbia, SC  2004         ford                 ranger   
13487     4488    gulfport / biloxi  2000       toyota                 tundra   
13971     2888      hattiesburg, MS  2000       toyota                4runner   
23574     2888          new orleans  2000       toyota                4runner   
25311     3995    okaloosa / walton  2007    chevrolet                 cobalt   
...        ...                  ...   ...          ...                    ...   
1903048  18995  dallas / fort worth  2010         ford   super duty f-350 srw   
1903322  29995  dallas / fort worth  2007          ram                   3500   
1907756  29999              el paso  2018         audi                     q7   
1913474  34999   mcallen / edinburg  2021         ford                 ranger   
1932118    650             richmond  2000         jeep  grand cherokee laredo   

         condition    cylinders    fuel  odometer odometer_range title_status  \
6485     excellent  6 cylinders     gas         0            NaN        clean   
13487         good  8 cylinders     gas         0            NaN        clean   
13971         good  6 cylinders     gas         0            NaN        clean   
23574         good  6 cylinders     gas         0            NaN        clean   
25311    excellent  4 cylinders     gas         0            NaN        clean   
...            ...          ...     ...       ...            ...          ...   
1903048  excellent  8 cylinders  diesel         0            NaN        clean   
1903322  excellent  6 cylinders  diesel         0            NaN        clean   
1907756  excellent  6 cylinders     gas         0            NaN        clean   
1913474        new  4 cylinders     gas         0            NaN        clean   
1932118       fair  6 cylinders     gas         0            NaN        clean   

        transmission drive       size    type paint_color  \
6485       automatic   4wd  full-size   truck        grey   
13487      automatic   4wd  full-size   truck       black   
13971      automatic   rwd  full-size     SUV       white   
23574      automatic   rwd  full-size     SUV       white   
25311      automatic   fwd    compact   coupe      yellow   
...              ...   ...        ...     ...         ...   
1903048    automatic   4wd  full-size  pickup       white   
1903322    automatic   4wd  full-size  pickup       white   
1907756    automatic   rwd  full-size     SUV       white   
1913474    automatic   rwd   mid-size   truck        grey   
1932118    automatic   4wd   mid-size     SUV      silver   

                     posting_date  
6485                          NaN  
13487                         NaN  
13971                         NaN  
23574                         NaN  
25311                         NaN  
...                           ...  
1903048  2021-05-04T16:15:16-0500  
1903322  2021-05-04T11:18:32-0500  
1907756  2021-04-08T14:59:57-0600  
1913474  2021-04-12T12:05:24-0500  
1932118  2021-04-06T19:45:51-0400  

[518 rows x 17 columns]

In [44]:
"""The rows will null values for 'odometer_range' are the rows which an odometer reading of 0. 
We categorize these rows as '0-12k' as well."""
raw_data_p11['odometer_range'] = raw_data_p11['odometer_range'].fillna('0-12k')
raw_data_p11.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 194861 entries, 102 to 1961146
Data columns (total 17 columns):
 #   Column          Non-Null Count   Dtype   
---  ------          --------------   -----   
 0   price           194861 non-null  int64   
 1   region          194861 non-null  object  
 2   year            194861 non-null  object  
 3   manufacturer    194861 non-null  object  
 4   model           194861 non-null  object  
 5   condition       194861 non-null  object  
 6   cylinders       194861 non-null  object  
 7   fuel            194861 non-null  object  
 8   odometer        194861 non-null  float64 
 9   odometer_range  194861 non-null  category
 10  title_status    194861 non-null  object  
 11  transmission    194861 non-null  object  
 12  drive           194861 non-null  object  
 13  size            194861 non-null  object  
 14  type            194861 non-null  object  
 15  paint_color     194861 non-null  object  
 16  posting_date    81409 non-null   ob

In [45]:
raw_data_p11['odometer_range'].nunique()

31

In [46]:
#Remove the column 'odometer'
raw_data_p11 = raw_data_p11.drop(['odometer'], axis=1)
raw_data_p11.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 194861 entries, 102 to 1961146
Data columns (total 16 columns):
 #   Column          Non-Null Count   Dtype   
---  ------          --------------   -----   
 0   price           194861 non-null  int64   
 1   region          194861 non-null  object  
 2   year            194861 non-null  object  
 3   manufacturer    194861 non-null  object  
 4   model           194861 non-null  object  
 5   condition       194861 non-null  object  
 6   cylinders       194861 non-null  object  
 7   fuel            194861 non-null  object  
 8   odometer_range  194861 non-null  category
 9   title_status    194861 non-null  object  
 10  transmission    194861 non-null  object  
 11  drive           194861 non-null  object  
 12  size            194861 non-null  object  
 13  type            194861 non-null  object  
 14  paint_color     194861 non-null  object  
 15  posting_date    81409 non-null   object  
dtypes: category(1), int64(1), object(14

#### P12 = Change the 'price' column to a categorical variable

In [47]:
#check the minimum and maximum values
raw_data_p11.describe()

price
count 194861
mean   12277
std    10331
min      505
25%     4995
50%     8995
75%    16500
max   131500

In [48]:
#create a list of bins for the ranges
price_bins = []
for i in range(28):
    price_bins.append(5000*i)
price_bins

[0,
 5000,
 10000,
 15000,
 20000,
 25000,
 30000,
 35000,
 40000,
 45000,
 50000,
 55000,
 60000,
 65000,
 70000,
 75000,
 80000,
 85000,
 90000,
 95000,
 100000,
 105000,
 110000,
 115000,
 120000,
 125000,
 130000,
 135000]

In [49]:
#create a list of labels for the ranges
price_labels = ['0-5k', '5k-10k', '10k-15k', '15k-20k', '20k-25k', '25k-30k', '30k-35k', '35k-40k', '40k-45k', '45k-50k',
               '50k-55k', '55k-60k', '60k-65k', '65k-70k', '70k-75k', '75k-80k', '80k-85k', '85k-90k', '90k-95k', '95k-100k',
               '100k-105k', '105k-110k', '110k-115k', '115k-120k', '120k-125k', '125k-130k', '130k-135k']

In [50]:
len(price_labels)

27

In [51]:
#Create the new column 'price_range'
raw_data_p12 = raw_data_p11.copy()
price_category = pd.cut(raw_data_p12.price,bins=price_bins,labels=price_labels)
raw_data_p12.insert(1,'price_range',price_category)
raw_data_p12

price price_range      region  year manufacturer  \
102      18990     15k-20k  albany, NY  2005    chevrolet   
104      19990     15k-20k  albany, NY  2008    chevrolet   
105      10990     10k-15k  albany, NY  2008    chevrolet   
107      17990     15k-20k  albany, NY  2005    chevrolet   
108      23990     20k-25k  albany, NY  2008    chevrolet   
...        ...         ...         ...   ...          ...   
1961120  23995     20k-25k     wyoming  2013       toyota   
1961121  15999     15k-20k     wyoming  2000          ram   
1961122  24995     20k-25k     wyoming  2014          gmc   
1961126  12995     10k-15k     wyoming  2016    chevrolet   
1961146  18000     15k-20k     wyoming  2005    chevrolet   

                         model  condition    cylinders    fuel odometer_range  \
102           silverado 2500hd  excellent  8 cylinders  diesel      216k-228k   
104           silverado 2500hd  excellent  8 cylinders  diesel      168k-180k   
105             silverado 1500  excellent  8 cylinders     gas      168k-180k   
107           silverado 2500hd  excellent  8 cylinders  diesel      108k-120k   
108           silverado 2500hd  excellent  8 cylinders  diesel      132k-144k   
...                        ...        ...          ...     ...            ...   
1961120                 tundra       good  8 cylinders     gas      108k-120k   
1961121                   3500       good  6 cylinders     gas      192k-204k   
1961122                 sierra       good  8 cylinders     gas      120k-132k   
1961126               cruze lt   like new  4 cylinders     gas        60k-72k   
1961146  silverado 1500 lt 4x4  excellent  8 cylinders     gas      120k-132k   

        title_status transmission drive       size    type paint_color  \
102            clean    automatic   4wd  full-size  pickup        blue   
104            clean    automatic   4wd  full-size   truck        blue   
105            clean    automatic   4wd  full-size  pickup      purple   
107            clean    automatic   4wd  full-size   truck        blue   
108            clean    automatic   4wd  full-size  pickup        blue   
...              ...          ...   ...        ...     ...         ...   
1961120        clean    automatic   4wd  full-size   truck       black   
1961121        clean    automatic   4wd  full-size   truck       white   
1961122        clean    automatic   4wd  full-size   truck       white   
1961126        clean    automatic   fwd    compact   sedan      silver   
1961146         lien    automatic   4wd  full-size   truck        blue   

                     posting_date  
102                           NaN  
104                           NaN  
105                           NaN  
107                           NaN  
108                           NaN  
...                           ...  
1961120  2021-04-07T14:42:51-0600  
1961121  2021-04-07T14:42:50-0600  
1961122  2021-04-07T14:42:50-0600  
1961126  2021-04-07T11:58:09-0600  
1961146  2021-04-06T21:04:03-0600  

[194861 rows x 17 columns]

In [52]:
#Remove the column 'price'
raw_data_p12 = raw_data_p12.drop(['price'], axis=1)
raw_data_p12.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 194861 entries, 102 to 1961146
Data columns (total 16 columns):
 #   Column          Non-Null Count   Dtype   
---  ------          --------------   -----   
 0   price_range     194861 non-null  category
 1   region          194861 non-null  object  
 2   year            194861 non-null  object  
 3   manufacturer    194861 non-null  object  
 4   model           194861 non-null  object  
 5   condition       194861 non-null  object  
 6   cylinders       194861 non-null  object  
 7   fuel            194861 non-null  object  
 8   odometer_range  194861 non-null  category
 9   title_status    194861 non-null  object  
 10  transmission    194861 non-null  object  
 11  drive           194861 non-null  object  
 12  size            194861 non-null  object  
 13  type            194861 non-null  object  
 14  paint_color     194861 non-null  object  
 15  posting_date    81409 non-null   object  
dtypes: category(2), object(14)
memory u

In [53]:
#Export this to a csv file for exploratory data analysis
raw_data_p12.to_csv('preprocessed_data_eda.csv', index=False)

#### P13 = One-hot encode the categorical variables

In [54]:
independent_variables = raw_data_p12.iloc[:,1:-1].columns.tolist()
independent_variables

['region',
 'year',
 'manufacturer',
 'model',
 'condition',
 'cylinders',
 'fuel',
 'odometer_range',
 'title_status',
 'transmission',
 'drive',
 'size',
 'type',
 'paint_color']

In [55]:
for c in independent_variables:
    print (c, raw_data_p12[c].nunique())

region 353
year 30
manufacturer 32
model 377
condition 6
cylinders 8
fuel 5
odometer_range 31
title_status 6
transmission 3
drive 3
size 4
type 13
paint_color 12


In [56]:
#one-hot encode the categorical variables
from sklearn.preprocessing import OneHotEncoder

ohe = OneHotEncoder(handle_unknown='ignore')
encoded_data = pd.DataFrame(ohe.fit_transform(raw_data_p12[independent_variables]).toarray())
encoded_data.columns = ohe.get_feature_names(independent_variables)
encoded_data

region_SF bay area  region_akron / canton  region_albany  \
0                        0                      0              0   
1                        0                      0              0   
2                        0                      0              0   
3                        0                      0              0   
4                        0                      0              0   
...                    ...                    ...            ...   
194856                   0                      0              0   
194857                   0                      0              0   
194858                   0                      0              0   
194859                   0                      0              0   
194860                   0                      0              0   

        region_albany, NY  region_albuquerque  region_amarillo  \
0                       1                   0                0   
1                       1                   0                0   
2                       1                   0                0   
3                       1                   0                0   
4                       1                   0                0   
...                   ...                 ...              ...   
194856                  0                   0                0   
194857                  0                   0                0   
194858                  0                   0                0   
194859                  0                   0                0   
194860                  0                   0                0   

        region_amarillo, TX  region_anchorage / mat-su  region_ann arbor  \
0                         0                          0                 0   
1                         0                          0                 0   
2                         0                          0                 0   
3                         0                          0                 0   
4                         0                          0                 0   
...                     ...                        ...               ...   
194856                    0                          0                 0   
194857                    0                          0                 0   
194858                    0                          0                 0   
194859                    0                          0                 0   
194860                    0                          0                 0   

        region_ann arbor, MI  ...  paint_color_brown  paint_color_custom  \
0                          0  ...                  0                   0   
1                          0  ...                  0                   0   
2                          0  ...                  0                   0   
3                          0  ...                  0                   0   
4                          0  ...                  0                   0   
...                      ...  ...                ...                 ...   
194856                     0  ...                  0                   0   
194857                     0  ...                  0                   0   
194858                     0  ...                  0                   0   
194859                     0  ...                  0                   0   
194860                     0  ...                  0                   0   

        paint_color_green  paint_color_grey  paint_color_orange  \
0                       0                 0                   0   
1                       0                 0                   0   
2                       0                 0                   0   
3                       0                 0                   0   
4                       0                 0                   0   
...                   ...               ...                 ...   
194856                  0                 0                   0   
194857                  0           

In [57]:
#remove the original columns for the categorical variables in the dataframe
raw_data_p13 = raw_data_p12.drop(independent_variables,axis=1)
raw_data_p13

price_range              posting_date
102         15k-20k                       NaN
104         15k-20k                       NaN
105         10k-15k                       NaN
107         15k-20k                       NaN
108         20k-25k                       NaN
...             ...                       ...
1961120     20k-25k  2021-04-07T14:42:51-0600
1961121     15k-20k  2021-04-07T14:42:50-0600
1961122     20k-25k  2021-04-07T14:42:50-0600
1961126     10k-15k  2021-04-07T11:58:09-0600
1961146     15k-20k  2021-04-06T21:04:03-0600

[194861 rows x 2 columns]

In [58]:
#add the encoded categorical variables to the dataframe
encoded_data.reset_index(drop=True, inplace=True)
raw_data_p13.reset_index(drop=True, inplace=True)
combined_data = pd.concat([raw_data_p13, encoded_data],axis=1)
combined_data

price_range              posting_date  region_SF bay area  \
0          15k-20k                       NaN                   0   
1          15k-20k                       NaN                   0   
2          10k-15k                       NaN                   0   
3          15k-20k                       NaN                   0   
4          20k-25k                       NaN                   0   
...            ...                       ...                 ...   
194856     20k-25k  2021-04-07T14:42:51-0600                   0   
194857     15k-20k  2021-04-07T14:42:50-0600                   0   
194858     20k-25k  2021-04-07T14:42:50-0600                   0   
194859     10k-15k  2021-04-07T11:58:09-0600                   0   
194860     15k-20k  2021-04-06T21:04:03-0600                   0   

        region_akron / canton  region_albany  region_albany, NY  \
0                           0              0                  1   
1                           0              0                  1   
2                           0              0                  1   
3                           0              0                  1   
4                           0              0                  1   
...                       ...            ...                ...   
194856                      0              0                  0   
194857                      0              0                  0   
194858                      0              0                  0   
194859                      0              0                  0   
194860                      0              0                  0   

        region_albuquerque  region_amarillo  region_amarillo, TX  \
0                        0                0                    0   
1                        0                0                    0   
2                        0                0                    0   
3                        0                0                    0   
4                        0                0                    0   
...                    ...              ...                  ...   
194856                   0                0                    0   
194857                   0                0                    0   
194858                   0                0                    0   
194859                   0                0                    0   
194860                   0                0                    0   

        region_anchorage / mat-su  ...  paint_color_brown  paint_color_custom  \
0                               0  ...                  0                   0   
1                               0  ...                  0                   0   
2                               0  ...                  0                   0   
3                               0  ...                  0                   0   
4                               0  ...                  0                   0   
...                           ...  ...                ...                 ...   
194856                          0  ...                  0                   0   
194857                          0  ...                  0                   0   
194858                          0  ...                  0                   0   
194859                          0  ...                  0                   0   
194860                          0  ...                  0                   0   

        paint_color_green  paint_color_grey  paint_color_orange  \
0                       0                 0                   0   
1                       0                 0                   0   
2                       0                 0                   0   
3                       0                 0                   0   
4                       0                 0                   0   
...                   ...               ...                 ...   
194856                  0                 0                   0   
194857                  0                 0                

In [59]:
#combined_data.to_csv('preprocessed_data.csv', index=False)